components/splitter/splitter.component.ts

```diff
@@ -0,0 +1,382 @@
+    const innerSizes = this.innerSizes();
+    const sizes = this.panelProps().map((panel, index) => {
+      const innerSize = innerSizes[index];
+      const size = panel.size ?? innerSize;
+      const hasSize = panel.size !== undefined;
+
+      // Calculate the percentage size of each panel.
+      let percentage: number | undefined;
+      if (isPercent(size)) {
+        percentage = getPercentValue(size);
+      } else if (size || size === 0) {
+        const num = Number(size);
+        if (!isNaN(num)) {
+          percentage = num / containerSize;
+        }
+      } else {
+        percentage = undefined;
+        emptyCount++;
+      }
+
+      // Calculate the min and max percentage size of each panel.
+      const minSize = panel.min;
+      const maxSize = panel.max;
+      const postPercentMinSize = isPercent(minSize) ? getPercentValue(minSize) : (minSize || 0) / containerSize;
+      const postPercentMaxSize = isPercent(maxSize)
+        ? getPercentValue(maxSize)
+        : (maxSize || containerSize) / containerSize;
+
+      return {
+        innerSize,
+        size,
+        hasSize,
+        percentage,
+        min: minSize,
+        max: maxSize,
+        postPercentMinSize,
+        postPercentMaxSize
+      } as PanelSize;
+    });
+
+    const totalPercentage = sizes.reduce((acc, { percentage }) => acc + (percentage ?? 0), 0);
+
+    if (totalPercentage > 1 && !emptyCount) {
+      // If total percentage is larger than 1, we will scale it down.
+      const scale = 1 / totalPercentage;
+      sizes.forEach(size => {
+        size.percentage = size.percentage === undefined ? 0 : size.percentage * scale;
+      });
+    } else {
+      // If total percentage is smaller than 1, we will fill the rest.
+      const averagePercentage = (1 - totalPercentage) / emptyCount;
+      sizes.forEach(size => {
+        size.percentage = size.percentage === undefined ? averagePercentage : size.percentage;
+      });
+    }
+
+    sizes.forEach(size => {
+      size.postPxSize = size.percentage * containerSize;
+      size.size = containerSize ? coerceCssPixelValue(size.postPxSize) : size.size;
+    });
+
+    return sizes;
+  });
+
+  /** ------------------ Resize ------------------ */
+  /**
+   * The index of the panel that is being resized.
+   * Mark the moving splitter bar as activated to show the dragging effect
+   * even if the mouse is outside the splitter container.
+   */
+  readonly movingIndex = signal<number | null>(null);
+  /**
+   * The offset of preview position (lazy mode) when dragging the splitter bar.
+   * Constrained by the min and max size of the target panel.
+   */
+  readonly constrainedOffset = signal<number>(0);
+  /**
+   * Handle the resize start event for the specified panel.
+   * @param index The index of the panel.
+   * @param startPos The start position of the resize event.
+   */
+  protected startResize(index: number, startPos: [x: number, y: number]): void {
```

Comment:
Unless one debugs this function (but still) it's very hard to understand what it does.  
Maybe a short paragraph explaining the theory behind it would be beneficial to other maintainers.